pallets/proofs-dealer/src/utils.rs

```diff
@@ -250,25 +271,266 @@ where
+        TickToChallengesSeed::<T>::set(challenges_ticker, Some(seed));
+        weight.consume(T::DbWeight::get().reads_writes(0, 1));
+
+        // Remove the oldest challenge seed stored, to clean up the storage.
+        let tick_to_remove = challenges_ticker.checked_sub(&ChallengeHistoryLengthFor::<T>::get());
+        if let Some(tick_to_remove) = tick_to_remove {
+            TickToChallengesSeed::<T>::remove(tick_to_remove);
+            weight.consume(T::DbWeight::get().reads_writes(0, 1));
+        }
+
+        // Emit new challenge seed event.
+        Self::deposit_event(Event::NewChallengeSeed {
+            challenges_ticker,
+            seed,
+        });
+
+        // Calculate if this is a checkpoint challenge round.
+        let last_checkpoint_tick = LastCheckpointTick::<T>::get();
+        // This hook does not return an error, and it cannot fail, that's why we use `saturating_add`.
+        let next_checkpoint_tick =
+            last_checkpoint_tick.saturating_add(T::CheckpointChallengePeriod::get());
+        if challenges_ticker == next_checkpoint_tick {
+            // This is a checkpoint challenge round, so we also generate new checkpoint challenges.
+            Self::do_new_checkpoint_challenge_round(challenges_ticker, weight);
+        }
+        weight.consume(T::DbWeight::get().reads_writes(2, 0));
+
+        // If there are providers left in `ChallengeTickToChallengedProviders` for this tick,
+        // they are marked as slashable.
+        let slashable_providers: Vec<_> =
+            ChallengeTickToChallengedProviders::<T>::iter_key_prefix(challenges_ticker).collect();
+        let slashable_providers_len = slashable_providers.len();
+        weight.consume(T::DbWeight::get().reads_writes(slashable_providers_len as u64, 0));
+        for provider in slashable_providers {
+            // Mark this provider as slashable.
+            SlashableProviders::<T>::set(provider, Some(()));
+            weight.consume(T::DbWeight::get().reads_writes(0, 1));
+
+            // Remove this tick entry from `ChallengeTickToChallengedProviders`.
+            ChallengeTickToChallengedProviders::<T>::remove(challenges_ticker, provider);
+            weight.consume(T::DbWeight::get().reads_writes(0, 1));
+
+            // Get the stake for this Provider, to know its challenge period.
+            // If a submitter is a registered Provider, it must have a stake, so there shouldn't be an error.
+            let stake = match ProvidersPalletFor::<T>::get_stake(provider) {
+                Some(stake) => stake,
+                // But to avoid panics, in the odd case of a Provider not being registered, we
+                // arbitrarily set the stake to be that which would result in 10 ticks of challenge period.
+                None => {
+                    weight.consume(T::DbWeight::get().reads_writes(1, 0));
+                    StakeToChallengePeriodFor::<T>::get() * 10u32.into()
```

Comment:
Maybe, but there's nothing this hook can do about it as it has no control over an implementation of a Providers pallet. we could maybe add an event to "flag" this. Also, we can't error out in the hook. But there is something I can do to improve, which is instead of using an arbitrary number 10, I use `CheckpointChallengesPeriod`.